book/content/reproducible_environments/09/case-study-branchLSTM.md

```diff
@@ -0,0 +1,110 @@
+When I compared these features across machines, I noticed something rather odd - the contents of corresponding elements were identical, apart from those which stored the length (to be more precise, the character count) of some of the tweets.
+
+Judging the length of a tweet seemed like a very simple task - just count the number of characters.
+So why were the counts different? Further investigation showed that the mismatched counts occurred in tweets containing emoticons, which led me to look into how these characters were represented in Unicode.
+
+Unicode characters are stored as one or more code points, with the specification dependent on whether the version of Python used is a wide or narrow build.
+Most of these terms were unfamiliar to me, so I’d recommend [this blog](https://www.b-list.org/weblog/2017/sep/05/how-python-does-unicode/) post if you’d like a detailed rundown.
+On a wide build, each code point is stored in four bytes, while on a narrow build only two bytes are required.
+This is fine for most characters, but code points that fall outside of the range that can be represented by two bytes are instead coded as surrogate pairs of two code points.
+It was differences in the build type of Python that had caused the differences in tweet length - on my MacBook Pro, emoticons were calculated to have length two as they were represented using surrogate pairs, while the length was calculated as expected, as a single character, on Linux systems (this meant that a consistent length of one had been used in Elena’s paper).
+As in the previous example about file load order, this is something that wasn't an issue when Elena originally performed her analysis as she worked on only one machine, but moving the code out of its original computational environment exposed these differences.
+
+To ensure the length of the tweet was calculated consistently, I replaced the pairs of Unicode characters with placeholder letters using a regular expression.
+Using the `len()` command to obtain the tweet length then gave consistent results on all machines.
+
+Note: As of Python 3.3 and later, there is a single build rather than wide or narrow options and Unicode is handled in the same way on all systems.
+This project used Python 2.7.
+
+## Returning to the package versions
+
+With the preprocessing stage now giving consistent representations of the data across all machines, it was time to look at the influences of the package versions on the output.
+At the time Elena was developing this project, neither of the Theano or Lasagne packages had reached version 1.0.
+In [semantic versioning](https://semver.org/), such 0.x.x version numbers indicate that the package may undergo large changes with each release, and so getting the version numbers correct was important if we were to ensure the output of the model was as intended.
+
+I knew roughly when Elena had installed the packages, so I started with the versions released around that time.
```

Comment:
I'm interested in this process - was it very manual or is there a tool to do this? Did it involve checking each package page on Pypi?

Reply:
I focussed on just Theano and Lasagne in this section as they both affected the model and were undergoing fairly rapid development at the time Elena did the work - with just two packages to consider, a manual check wasn't too time-consuming. I've added another sentence explaining and linking out to the releases.